<!DOCTYPE html><html lang="en" mode="dark" ><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="day-prompt" content="days ago"><meta name="hour-prompt" content="hours ago"><meta name="minute-prompt" content="minutes ago"><meta name="justnow-prompt" content="just now"><meta name="generator" content="Jekyll v4.2.1" /><meta property="og:title" content="Playing with bioinformatics I" /><meta name="author" content="Lucas M. Taniguti" /><meta property="og:locale" content="en" /><meta name="description" content="Bioinformatics is known to have plenty of file formats. Every time a discussion about file formats starts, I remember this comic." /><meta property="og:description" content="Bioinformatics is known to have plenty of file formats. Every time a discussion about file formats starts, I remember this comic." /><link rel="canonical" href="https://lmtani.github.io/posts/playing-with-bioinformatics-I/" /><meta property="og:url" content="https://lmtani.github.io/posts/playing-with-bioinformatics-I/" /><meta property="og:site_name" content="lmtani" /><meta property="og:type" content="article" /><meta property="article:published_time" content="2021-10-15T00:33:00-03:00" /><meta name="twitter:card" content="summary" /><meta property="twitter:title" content="Playing with bioinformatics I" /><meta name="twitter:site" content="@ltaniguti" /><meta name="twitter:creator" content="@Lucas M. Taniguti" /><meta name="google-site-verification" content="google_meta_tag_verification" /> <script type="application/ld+json"> {"author":{"@type":"Person","name":"Lucas M. Taniguti"},"headline":"Playing with bioinformatics I","dateModified":"2021-10-24T09:57:35-03:00","datePublished":"2021-10-15T00:33:00-03:00","url":"https://lmtani.github.io/posts/playing-with-bioinformatics-I/","description":"Bioinformatics is known to have plenty of file formats. Every time a discussion about file formats starts, I remember this comic.","@type":"BlogPosting","mainEntityOfPage":{"@type":"WebPage","@id":"https://lmtani.github.io/posts/playing-with-bioinformatics-I/"},"@context":"https://schema.org"}</script><title>Playing with bioinformatics I | lmtani</title><link rel="apple-touch-icon" sizes="180x180" href="/assets/img/favicons/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="/assets/img/favicons/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="/assets/img/favicons/favicon-16x16.png"><link rel="manifest" href="/assets/img/favicons/site.webmanifest"><link rel="shortcut icon" href="/assets/img/favicons/favicon.ico"><meta name="apple-mobile-web-app-title" content="lmtani"><meta name="application-name" content="lmtani"><meta name="msapplication-TileColor" content="#da532c"><meta name="msapplication-config" content="/assets/img/favicons/browserconfig.xml"><meta name="theme-color" content="#ffffff"><link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="anonymous"><link rel="dns-prefetch" href="https://fonts.gstatic.com"><link rel="preconnect" href="https://www.google-analytics.com" crossorigin="use-credentials"><link rel="dns-prefetch" href="https://www.google-analytics.com"><link rel="preconnect" href="https://www.googletagmanager.com" crossorigin="anonymous"><link rel="dns-prefetch" href="https://www.googletagmanager.com"><link rel="preconnect" href="https://cdn.jsdelivr.net"><link rel="dns-prefetch" href="https://cdn.jsdelivr.net"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.0.0/dist/css/bootstrap.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.11.2/css/all.min.css"><link rel="stylesheet" href="/assets/css/style.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@1.0.1/dist/bootstrap-toc.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/magnific-popup@1.1.0/dist/magnific-popup.min.css"> <script src="https://cdn.jsdelivr.net/npm/jquery@3/dist/jquery.min.js"></script><body data-spy="scroll" data-target="#toc"><div id="sidebar" class="d-flex flex-column align-items-end" lang="en"><div class="profile-wrapper text-center"><div id="avatar"> <a href="/" alt="avatar" class="mx-auto"> <img src="https://avatars.githubusercontent.com/u/12699242?s=400&u=8a10e268f99e49e1ce2ae70fb140e698a7655e86&v=4" alt="avatar" onerror="this.style.display='none'"> </a></div><div class="site-title mt-3"> <a href="/">lmtani</a></div><div class="site-subtitle font-italic">A personal blog with things that interests me</div></div><ul class="w-100"><li class="nav-item"> <a href="/" class="nav-link"> <i class="fa-fw fas fa-home ml-xl-3 mr-xl-3 unloaded"></i> <span>HOME</span> </a><li class="nav-item"> <a href="/categories/" class="nav-link"> <i class="fa-fw fas fa-stream ml-xl-3 mr-xl-3 unloaded"></i> <span>CATEGORIES</span> </a><li class="nav-item"> <a href="/tags/" class="nav-link"> <i class="fa-fw fas fa-tag ml-xl-3 mr-xl-3 unloaded"></i> <span>TAGS</span> </a><li class="nav-item"> <a href="/archives/" class="nav-link"> <i class="fa-fw fas fa-archive ml-xl-3 mr-xl-3 unloaded"></i> <span>ARCHIVES</span> </a><li class="nav-item"> <a href="/about/" class="nav-link"> <i class="fa-fw fas fa-info-circle ml-xl-3 mr-xl-3 unloaded"></i> <span>ABOUT</span> </a></ul><div class="sidebar-bottom mt-auto d-flex flex-wrap justify-content-center align-items-center"> <a href="https://github.com/lmtani" aria-label="github" target="_blank" rel="noopener"> <i class="fab fa-github"></i> </a> <a href="https://twitter.com/ltaniguti" aria-label="twitter" target="_blank" rel="noopener"> <i class="fab fa-twitter"></i> </a> <a href=" javascript:location.href = 'mailto:' + ['0l0y34v5r','relay.firefox.com'].join('@')" aria-label="email" > <i class="fas fa-envelope"></i> </a> <a href="/feed.xml" aria-label="rss" > <i class="fas fa-rss"></i> </a></div></div><div id="topbar-wrapper" class="row justify-content-center topbar-down"><div id="topbar" class="col-11 d-flex h-100 align-items-center justify-content-between"> <span id="breadcrumb"> <span> <a href="/"> Home </a> </span> <span>Playing with bioinformatics I</span> </span> <i id="sidebar-trigger" class="fas fa-bars fa-fw"></i><div id="topbar-title"> Post</div><i id="search-trigger" class="fas fa-search fa-fw"></i> <span id="search-wrapper" class="align-items-center"> <i class="fas fa-search fa-fw"></i> <input class="form-control" id="search-input" type="search" aria-label="search" autocomplete="off" placeholder="Search..."> <i class="fa fa-times-circle fa-fw" id="search-cleaner"></i> </span> <span id="search-cancel" >Cancel</span></div></div><div id="main-wrapper"><div id="main"><div class="row"><div id="post-wrapper" class="col-12 col-lg-11 col-xl-8"><div class="post pl-1 pr-1 pl-sm-2 pr-sm-2 pl-md-4 pr-md-4"><h1 data-toc-skip>Playing with bioinformatics I</h1><div class="post-meta text-muted d-flex flex-column"><div> <span class="semi-bold"> Lucas M. Taniguti </span> on <span class="timeago " data-toggle="tooltip" data-placement="bottom" title="Fri, Oct 15, 2021, 12:33 AM -0300" >Oct 15<i class="unloaded">2021-10-15T00:33:00-03:00</i> </span></div><div> <span> Updated <span class="timeago lastmod" data-toggle="tooltip" data-placement="bottom" title="Sun, Oct 24, 2021, 9:57 AM -0300" >Oct 24<i class="unloaded">2021-10-24T09:57:35-03:00</i> </span> </span> <span class="readtime" data-toggle="tooltip" data-placement="bottom" title="1179 words">6 min read</span></div></div><div class="post-content"><p>Bioinformatics is known to have plenty of file formats. Every time a discussion about file formats starts, I remember <a href="http://www.niso.org/sites/default/files/inline-images/How%20Standards%20Proliferate%20%281%29.png">this comic</a>.</p><p>I’ll guide you through some of these formats in this post, always paying attention to our precious Illumina reads contents. Here is our plan:</p><ol><li>Download public data to play with. We’ll use the sars-cov-2 genome, and Illumina reads from a whole-genome amplicon experiment.<li>Convert FASTQ files to unaligned BAM (uBAM), just to show how it works.<li>Transform uBAM into a single interleave FASTQ file.<li>Align FASTQs files (interleaved and paired) on the sars-cov-2 genome, creating a BAM file.<li>Convert BAM to CRAM and also to a lossy CRAM<li>Convert back from CRAM to pairs of FASTQs</ol><p>You can find the shell script we use here in this <a href="https://github.com/lmtani/lmtani.github.io/blob/wip-illumina-format-patterns/_code/playing-with-bioinformatics-files/playing-with-bioinformatics-files.sh">GitHub repository</a>.</p><h2 id="tools">Tools</h2><p>I’m assuming you are in a Unix-like system (ubuntu, fedora, WSL2, macOS, etc.)</p><ul><li><a href="https://broadinstitute.github.io/picard/">Picard</a> - To convert FASTQ ⇄ uBAM and SAM ⇄ BAM ⇄ CRAM<li><a href="http://www.htslib.org/">Samtools</a> - To create .fai index<li><a href="https://github.com/lh3/bwa">BWA</a> - To align reads to genome<li><a href="https://www.sanger.ac.uk/tool/crumble/">Crumble</a> - To try lossy compression on alignment file</ul><blockquote><p>See <a href="https://docs.conda.io/en/latest/miniconda.html">miniconda</a> or <a href="https://www.docker.com/">Docker</a> to easily get your environment ready. I plan to make a post about it soon.</p></blockquote><h2 id="download-public-dataset">Download public dataset</h2><p>Firstly, go to a directory you like. For example, <code class="language-plaintext highlighter-rouge">/tmp/bioinfo</code> if you don’t want to persist files from this post after reboot your system.</p><p>We’ll use two repositories to retrieve some data for our experiment:</p><ul><li><a href="https://www.ncbi.nlm.nih.gov/sra">Sequence Read Archive (SRA)</a>: stores raw data from NGS sequencing experiments.<li>FTP from <a href="https://www.ncbi.nlm.nih.gov/genome">NCBI/genomes</a>: organizes information on genomes.</ul><h3 id="raw-sequencing-data">Raw sequencing data</h3><p>SRA organizes its data using unique identifiers, like this one under our <code class="language-plaintext highlighter-rouge">GENOME_SRA_ID</code> variable. If you want more information, visit <a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=ERR6335730">ERR6335730</a> page to inspect.</p><p>The tool <code class="language-plaintext highlighter-rouge">fastq-dump</code> is used to download reads and store them in two FASTQ files. One for R1 and the other for R2 reads.</p><script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Flmtani%2Flmtani.github.io%2Fblob%2Fwip-illumina-format-patterns%2F_code%2Fplaying-with-bioinformatics-files%2Fplaying-with-bioinformatics-files.sh%23L5-L7&amp;style=zenburn&amp;showLineNumbers=on&amp;showFileMeta=on&amp;showCopy=on"></script><h3 id="reference-sequence">Reference sequence</h3><p>Usually, reference sequences (or genomes) are stored in the <a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA format</a>.</p><p>In the code below, we download it with <code class="language-plaintext highlighter-rouge">wget</code> program, writing it as genome.fa.gz inside a directory. Then we decompress it with gunzip, and finally, we create a required index file (.fai) using samtools.</p><script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Flmtani%2Flmtani.github.io%2Fblob%2Fwip-illumina-format-patterns%2F_code%2Fplaying-with-bioinformatics-files%2Fplaying-with-bioinformatics-files.sh%23L9-L17&amp;style=zenburn&amp;showLineNumbers=on&amp;showFileMeta=on&amp;showCopy=on"></script><h2 id="transforming-reads">Transforming reads</h2><p>Here we’ll transform our R1 and R2 files into two different formats, just to illustrate our diversity of file formats.</p><h3 id="the-unaligned-bam-format">The unaligned BAM format</h3><p>In this step, we’ll convert R1 and R2 reads into a single <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">SAM-like</a> file named unaligned BAM (uBAM). The significant advantage I see in this format is that one can store metadata in its header, so you can have pieces of information like the sequencing run name, the library, the date of the experiment, the facility it was sequenced so on.</p><p>This is the format used by <a href="https://broadinstitute.github.io/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README/#input-requirements-and-expectations">Broad Institute in their workflows</a>.</p><p>Here is how to have it:</p><script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Flmtani%2Flmtani.github.io%2Fblob%2Fwip-illumina-format-patterns%2F_code%2Fplaying-with-bioinformatics-files%2Fplaying-with-bioinformatics-files.sh%23L20-L38&amp;style=zenburn&amp;showLineNumbers=on&amp;showFileMeta=on&amp;showCopy=on"></script><h3 id="the-interleaved-fastq-format">The interleaved FASTQ format</h3><p>This one is not as popular as the others, but I’ll mention it here just in case you see it out there. The two reads (R1 and R2) are in the same file and marked as /1 and /2 to inform their origin.</p><script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Flmtani%2Flmtani.github.io%2Fblob%2Fwip-illumina-format-patterns%2F_code%2Fplaying-with-bioinformatics-files%2Fplaying-with-bioinformatics-files.sh%23L40-L46&amp;style=zenburn&amp;showLineNumbers=on&amp;showFileMeta=on&amp;showCopy=on"></script><h2 id="align-to-reference-sequence">Align to reference sequence</h2><p>This step usually is the most intense in genotyping workflows. This step is responsible for identifying where is the best region to align each of your NGS reads. There is a lot of complexity in this step, but it’s out of the scope of this post. For now, you can consider the best region as the interval where your sequencing read and the reference sequence are more similar.</p><p>For comparison, we will align the pair of FASTQs (R1 and R2) and also the interleaved one.</p><script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Flmtani%2Flmtani.github.io%2Fblob%2Fwip-illumina-format-patterns%2F_code%2Fplaying-with-bioinformatics-files%2Fplaying-with-bioinformatics-files.sh%23L52-L61&amp;style=zenburn&amp;showLineNumbers=on&amp;showFileMeta=on&amp;showCopy=on"></script><p>It creates two alignment files named BAM. Remember that whenever you see SAM/BAM/CRAM files, they essentially store the same information: how your reads align to the reference sequence. SAM is not compressed, and CRAM is smaller in file size. Note that you’ll need to use your reference sequence to restore information from CRAM files.</p><p>Now that we have our BAM files, we can calculate some statistics using Picard. The goal in using this is to see that both files contain essentially the same data. Minor differences are expected.</p><script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Flmtani%2Flmtani.github.io%2Fblob%2Fwip-illumina-format-patterns%2F_code%2Fplaying-with-bioinformatics-files%2Fplaying-with-bioinformatics-files.sh%23L63-L68&amp;style=zenburn&amp;showLineNumbers=on&amp;showFileMeta=on&amp;showCopy=on"></script><p>See that both stats have very similar metrics. From now on, we will use only the BAM obtained from the alignment of the interleaved FASTQ.</p><div class="language-sh highlighter-rouge"><div class="code-header"> <span text-data=" Shell "><i class="fa-fw fas fa-code small"></i></span> <button aria-label="copy" title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
</pre><td class="rouge-code"><pre><span class="c"># zgrep to select only some lines and columns</span>
zgrep <span class="nt">-E</span> <span class="s2">"^PAIR|^CATEGORY"</span> algn-from-pairs.bam.stats | <span class="nb">cut</span> <span class="nt">-f</span> 1,2,3,4,5,6,7,8 | column <span class="nt">-t</span>
CATEGORY  TOTAL_READS  PF_READS  PCT_PF_READS  PF_NOISE_READS  PF_READS_ALIGNED  PCT_PF_READS_ALIGNED  PF_ALIGNED_BASES
PAIR      200000       200000    1             0               199302            0.99651               21604991

zgrep <span class="nt">-E</span> <span class="s2">"^PAIR|^CATEGORY"</span> algn-from-interleaved.bam.stats | <span class="nb">cut</span> <span class="nt">-f</span> 1,2,3,4,5,6,7,8 | column <span class="nt">-t</span>
CATEGORY  TOTAL_READS  PF_READS  PCT_PF_READS  PF_NOISE_READS  PF_READS_ALIGNED  PCT_PF_READS_ALIGNED  PF_ALIGNED_BASES
PAIR      200000       200000    1             0               199302            0.99651               21604976
</pre></table></code></div></div><h2 id="storing-as-cram-file">Storing as CRAM file</h2><p>CRAM is the most compressed alignment format, and because of this, it’s usually the adopted form for long-term storage. From what I see in human genome/exome data, it’s about 1/3 of the equivalent BAM file.</p><p>To create it:</p><script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Flmtani%2Flmtani.github.io%2Fblob%2Fwip-illumina-format-patterns%2F_code%2Fplaying-with-bioinformatics-files%2Fplaying-with-bioinformatics-files.sh%23L70-L73&amp;style=zenburn&amp;showLineNumbers=on&amp;showFileMeta=on&amp;showCopy=on"></script><blockquote><p>Remember: you will always need the reference sequence (FASTA) to decompress this format.</p></blockquote><h3 id="example-of-lossy-compression">Example of lossy compression</h3><p>Because of the amount of data that today’s sequencers can produce, storage is a challenge. You can use local or cloud-based systems, but the problem remains the same.</p><p>Crumble is a tool to perform controlled loss of information to improve the compression of alignments. Here we’ll execute it with the most “aggressive” parameters to show how much smaller an alignment could be if we throw away some of its information.</p><script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Flmtani%2Flmtani.github.io%2Fblob%2Fwip-illumina-format-patterns%2F_code%2Fplaying-with-bioinformatics-files%2Fplaying-with-bioinformatics-files.sh%23L75-L80&amp;style=zenburn&amp;showLineNumbers=on&amp;showFileMeta=on&amp;showCopy=on"></script><h3 id="alignment-file-size-comparison">Alignment file size comparison</h3><ul><li>BAM: 16M<li>CRAM: 9.1M<li>Lossy CRAM: 3.4M</ul><h2 id="recover-raw-reads-from-alignment">Recover raw reads from alignment</h2><p>Remember that you aligned your reads into a reference sequence, and this alignment is stored as SAM/BAM/CRAM. So, if you don’t have used any destructive parameter, you could retrieve the same read information from the alignment. This is important because you could delete the original sequences or put them in a coldline storage system if you prefer.</p><blockquote><p>There are a lot of cloud storage categories where it’s cheap to store, but you’ll pay more if you need to retrieve the data.</p></blockquote><p>Here we compare the content of one read. Headers are different, and because of this, we need to compare the sequence and quality only.</p><script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Flmtani%2Flmtani.github.io%2Fblob%2Fwip-illumina-format-patterns%2F_code%2Fplaying-with-bioinformatics-files%2Fplaying-with-bioinformatics-files.sh%23L82-L90&amp;style=zenburn&amp;showLineNumbers=on&amp;showFileMeta=on&amp;showCopy=on"></script><p>If you calculate the md5sum hash for these two files, you’ll see that they have the same content. You can also inspect it manually by opening in any text editor or using the <code class="language-plaintext highlighter-rouge">cat</code> command.</p><div class="language-sh highlighter-rouge"><div class="code-header"> <span text-data=" Shell "><i class="fa-fw fas fa-code small"></i></span> <button aria-label="copy" title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
</pre><td class="rouge-code"><pre><span class="c"># Original</span>
zgrep <span class="nt">-A</span> 3 <span class="nt">-w</span> <span class="s2">"@ERR6335730.9"</span> ERR6335730_1.fastq.gz | <span class="nb">grep</span> <span class="nt">-E</span> <span class="nt">-v</span> <span class="s2">"^</span><span class="se">\@</span><span class="s2">|^</span><span class="se">\+</span><span class="s2">"</span>
GAACTCACTTTCCATCCAACTTTTGTTGTTTTTGTGGTTAGAAGTAACACCCAAAAATGGATCATTACAAAATTGAAATTCACAGACTTTAATAACAACATTAGTAGCTGTCTCTTATACACATCTCCGAGCCCACGAGACGAGAATGGT
<span class="o">&gt;</span>3AAAFFFFFFFFFGD4FFGGGHHGHHGHHHHGGHH2GFGFCHDGGFGHGGGGHEEGFFHHHHEGHBHHHHGHHHFGBGHHEFGHFHHGHHGHFHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHGGGEGGGGGGEGGFGCDDHHFH

<span class="c"># Recovered from CRAM</span>
zgrep <span class="nt">-A</span> 3 <span class="nt">-w</span> <span class="s2">"@ERR6335730.9"</span> STEP_5/restored_R1.fq.gz | <span class="nb">grep</span> <span class="nt">-E</span> <span class="nt">-v</span> <span class="s2">"^</span><span class="se">\@</span><span class="s2">|^</span><span class="se">\+</span><span class="s2">"</span>
GAACTCACTTTCCATCCAACTTTTGTTGTTTTTGTGGTTAGAAGTAACACCCAAAAATGGATCATTACAAAATTGAAATTCACAGACTTTAATAACAACATTAGTAGCTGTCTCTTATACACATCTCCGAGCCCACGAGACGAGAATGGT
<span class="o">&gt;</span>3AAAFFFFFFFFFGD4FFGGGHHGHHGHHHHGGHH2GFGFCHDGGFGHGGGGHEEGFFHHHHEGHBHHHHGHHHFGBGHHEFGHFHHGHHGHFHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHGGGEGGGGGGEGGFGCDDHHFH

<span class="c"># If you want to know how it stands after crumble</span>
<span class="c"># you could revert its CRAM to FASTQ and inspect.</span>
zgrep <span class="nt">-A</span> 3 <span class="nt">-w</span> <span class="s2">"@ERR6335730.9"</span> from-crumble_1.fq.gz | <span class="nb">grep</span> <span class="nt">-E</span> <span class="nt">-v</span> <span class="s2">"^</span><span class="se">\@</span><span class="s2">|^</span><span class="se">\+</span><span class="s2">"</span>
GAACTCACTTTCCATCCAACTTTTGTTGTTTTTGTGGTTAGAAGTAACACCCAAAAATGGATCATTACAAAATTGAAATTCACAGACTTTAATAACAACATTAGTAGCTGTCTCTTATACACATCTCCGAGCCCACGAGACGAGAATGGT
88DDDDDDDDDDDDD<span class="o">&lt;&lt;</span><span class="no">GGGGGGGGGGGGGGGGGGG2EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE</span><span class="sh">
</span></pre></table></code></div></div><p>That’s it. Hope you find it helpful to get familiar with some common bioinformatics steps and file formats. Let me know if you have any questions or suggestions in the comments below.</p><h2 id="comments">Comments</h2><div id="disqus_thread" class="pt-2 pb-2"><p class="text-center text-muted small"> Comments powered by <a href="https://disqus.com/">Disqus</a>.</p></div><script type="text/javascript"> var disqus_config = function () { this.page.url = 'https://lmtani.github.io/posts/playing-with-bioinformatics-I/'; this.page.identifier = '/posts/playing-with-bioinformatics-I/'; }; /* Lazy loading */ var disqus_observer = new IntersectionObserver(function (entries) { if(entries[0].isIntersecting) { (function () { var d = document, s = d.createElement('script'); s.src = 'https://lmtani.disqus.com/embed.js'; s.setAttribute('data-timestamp', +new Date()); (d.head || d.body).appendChild(s); })(); disqus_observer.disconnect(); } }, { threshold: [0] }); disqus_observer.observe(document.querySelector('#disqus_thread')); /* Auto switch theme */ function reloadDisqus() { /* Disqus hasn't been loaded */ if (typeof DISQUS === "undefined") { return; } if (document.readyState == 'complete') { DISQUS.reset({ reload: true, config: disqus_config }); } } const modeToggle = document.querySelector(".mode-toggle"); if (modeToggle !== null) { modeToggle.addEventListener('click', reloadDisqus); window.matchMedia('(prefers-color-scheme: dark)').addEventListener('change', reloadDisqus); } </script></div><div class="post-tail-wrapper text-muted"><div class="post-meta mb-3"> <i class="far fa-folder-open fa-fw mr-1"></i> <a href='/categories/bioinformatics/'>Bioinformatics</a>, <a href='/categories/file-formats/'>File formats</a></div><div class="post-tags"> <i class="fa fa-tags fa-fw mr-1"></i> <a href="/tags/alignment/" class="post-tag no-text-decoration" >alignment</a> <a href="/tags/bwa/" class="post-tag no-text-decoration" >bwa</a> <a href="/tags/picard/" class="post-tag no-text-decoration" >picard</a> <a href="/tags/crumble/" class="post-tag no-text-decoration" >crumble</a> <a href="/tags/samtools/" class="post-tag no-text-decoration" >samtools</a></div><div class="post-tail-bottom d-flex justify-content-between align-items-center mt-3 pt-5 pb-2"><div class="license-wrapper"> This post is licensed under <a href="https://creativecommons.org/licenses/by/4.0/"> CC BY 4.0 </a> by the author.</div><div class="share-wrapper"> <span class="share-label text-muted mr-1">Share</span> <span class="share-icons"> <a href="https://twitter.com/intent/tweet?text=Playing with bioinformatics I - lmtani&url=https://lmtani.github.io/posts/playing-with-bioinformatics-I/" data-toggle="tooltip" data-placement="top" title="Twitter" target="_blank" rel="noopener" aria-label="Twitter"> <i class="fa-fw fab fa-twitter"></i> </a> <a href="https://www.facebook.com/sharer/sharer.php?title=Playing with bioinformatics I - lmtani&u=https://lmtani.github.io/posts/playing-with-bioinformatics-I/" data-toggle="tooltip" data-placement="top" title="Facebook" target="_blank" rel="noopener" aria-label="Facebook"> <i class="fa-fw fab fa-facebook-square"></i> </a> <a href="https://telegram.me/share?text=Playing with bioinformatics I - lmtani&url=https://lmtani.github.io/posts/playing-with-bioinformatics-I/" data-toggle="tooltip" data-placement="top" title="Telegram" target="_blank" rel="noopener" aria-label="Telegram"> <i class="fa-fw fab fa-telegram"></i> </a> <i id="copy-link" class="fa-fw fas fa-link small" data-toggle="tooltip" data-placement="top" title="Copy link" title-succeed="Link copied successfully!"> </i> </span></div></div></div></div></div><div id="panel-wrapper" class="col-xl-3 pl-2 text-muted topbar-down"><div class="access"><div id="access-lastmod" class="post"> <span>Recent Update</span><ul class="post-content pl-0 pb-1 ml-1 mt-2"><li><a href="/posts/blast-using-sequence-server/">BLAST using Sequence Server</a><li><a href="/posts/playing-with-bioinformatics-I/">Playing with bioinformatics I</a></ul></div><div id="access-tags"> <span>Trending Tags</span><div class="d-flex flex-wrap mt-3 mb-1 mr-3"> <a class="post-tag" href="/tags/alignment/">alignment</a> <a class="post-tag" href="/tags/blast/">BLAST</a> <a class="post-tag" href="/tags/bwa/">bwa</a> <a class="post-tag" href="/tags/crumble/">crumble</a> <a class="post-tag" href="/tags/picard/">picard</a> <a class="post-tag" href="/tags/samtools/">samtools</a></div></div></div><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@1.0.1/dist/bootstrap-toc.min.js"></script><div id="toc-wrapper" class="pl-0 pr-4 mb-5"> <span class="pl-3 pt-2 mb-2">Contents</span><nav id="toc" data-toggle="toc"></nav></div></div></div><div class="row"><div class="col-12 col-lg-11 col-xl-8"><div id="post-extend-wrapper" class="pl-1 pr-1 pl-sm-2 pr-sm-2 pl-md-4 pr-md-4"><div id="related-posts" class="mt-5 mb-2 mb-sm-4"><h3 class="pt-2 mt-1 mb-4 ml-1" data-toc-skip>Further Reading</h3><div class="card-deck mb-4"><div class="card"> <a href="/posts/blast-using-sequence-server/"><div class="card-body"> <span class="timeago small" >Oct 6<i class="unloaded">2021-10-06T00:33:00-03:00</i> </span><h3 class="pt-0 mt-1 mb-3" data-toc-skip>BLAST using Sequence Server</h3><div class="text-muted small"><p> Introduction BLAST is one of the most used bioinformatics tools. It allows researchers to quickly find regions of similarity between a sequence of interest and a database with thousands of other s...</p></div></div></a></div></div></div><div class="post-navigation d-flex justify-content-between"> <a href="/posts/blast-using-sequence-server/" class="btn btn-outline-primary" prompt="Older"><p>BLAST using Sequence Server</p></a> <span class="btn btn-outline-primary disabled" prompt="Newer"><p>-</p></span></div></div></div></div><footer class="d-flex w-100 justify-content-center"><div class="d-flex justify-content-between align-items-center"><div class="footer-left"><p class="mb-0"> © 2021 <a href="https://twitter.com/ltaniguti">Lucas M. Taniguti</a>. <span data-toggle="tooltip" data-placement="top" title="Except where otherwise noted, the blog posts on this site are licensed under the Creative Commons Attribution 4.0 International (CC BY 4.0) License by the author.">Some rights reserved.</span></p></div><div class="footer-right"><p class="mb-0"> Powered by <a href="https://jekyllrb.com" target="_blank" rel="noopener">Jekyll</a> with <a href="https://github.com/cotes2020/jekyll-theme-chirpy" target="_blank" rel="noopener">Chirpy</a> theme.</p></div></div></footer></div><div id="search-result-wrapper" class="d-flex justify-content-center unloaded"><div class="col-12 col-sm-11 post-content"><div id="search-hints"><h4 class="text-muted mb-4">Trending Tags</h4><a class="post-tag" href="/tags/alignment/">alignment</a> <a class="post-tag" href="/tags/blast/">BLAST</a> <a class="post-tag" href="/tags/bwa/">bwa</a> <a class="post-tag" href="/tags/crumble/">crumble</a> <a class="post-tag" href="/tags/picard/">picard</a> <a class="post-tag" href="/tags/samtools/">samtools</a></div><div id="search-results" class="d-flex flex-wrap justify-content-center text-muted mt-3"></div></div></div></div><div id="mask"></div><a id="back-to-top" href="#" aria-label="back-to-top" class="btn btn-lg btn-box-shadow" role="button"> <i class="fas fa-angle-up"></i> </a> <script src="https://cdn.jsdelivr.net/npm/simple-jekyll-search@1.10.0/dest/simple-jekyll-search.min.js"></script> <script> SimpleJekyllSearch({ searchInput: document.getElementById('search-input'), resultsContainer: document.getElementById('search-results'), json: '/assets/js/data/search.json', searchResultTemplate: '<div class="pl-1 pr-1 pl-sm-2 pr-sm-2 pl-lg-4 pr-lg-4 pl-xl-0 pr-xl-0"> <a href="https://lmtani.github.io{url}">{title}</a><div class="post-meta d-flex flex-column flex-sm-row text-muted mt-1 mb-1"> {categories} {tags}</div><p>{snippet}</p></div>', noResultsText: '<p class="mt-5">Oops! No result founds.</p>', templateMiddleware: function(prop, value, template) { if (prop === 'categories') { if (value === '') { return `${value}`; } else { return `<div class="mr-sm-4"><i class="far fa-folder fa-fw"></i>${value}</div>`; } } if (prop === 'tags') { if (value === '') { return `${value}`; } else { return `<div><i class="fa fa-tag fa-fw"></i>${value}</div>`; } } } }); </script> <script src="https://cdn.jsdelivr.net/combine/npm/lozad/dist/lozad.min.js,npm/magnific-popup@1/dist/jquery.magnific-popup.min.js,npm/clipboard@2/dist/clipboard.min.js"></script> <script defer src="/assets/js/dist/post.min.js"></script> <script src="https://cdn.jsdelivr.net/combine/npm/popper.js@1.16.1,npm/bootstrap@4/dist/js/bootstrap.min.js"></script> <script defer src="/app.js"></script> <script defer src="https://www.googletagmanager.com/gtag/js?id=G-D7Z4LDR8ZW"></script> <script> document.addEventListener("DOMContentLoaded", function(event) { window.dataLayer = window.dataLayer || []; function gtag(){dataLayer.push(arguments);} gtag('js', new Date()); gtag('config', 'G-D7Z4LDR8ZW'); }); </script>
